multi-pass-generation-prototype/record-example/06-output.py

```diff
@@ -0,0 +1,148 @@
+        """
+        builder = RustBufferBuilder()
+        try:
+            yield builder
+        except:
+            builder.discard()
+            raise
+
+    @contextlib.contextmanager
+    def consumeWithStream(self):
+        """Context-manager to consume a buffer using a RustBufferStream.
+
+        The RustBuffer will be freed once the context-manager exits, ensuring that we don't
+        leak it even if an error occurs.
+        """
+        try:
+            s = RustBufferStream(self)
+            yield s
+            if s.remaining() != 0:
+                raise RuntimeError("junk data left in buffer after consuming")
+        finally:
+            self.free()
+
+class RustBufferBuilder(object):
+    """Helper for structured writing of values into a RustBuffer."""
+
+    def __init__(self):
+        self.rbuf = RustBuffer.alloc(16)
+        self.rbuf.len = 0
+
+    def finalize(self):
+        rbuf = self.rbuf
+        self.rbuf = None
+        return rbuf
+
+    def discard(self):
+        if self.rbuf is not None:
+            rbuf = self.finalize()
+            rbuf.free()
+
+    @contextlib.contextmanager
+    def _reserve(self, numBytes):
+        if self.rbuf.len + numBytes > self.rbuf.capacity:
+            self.rbuf = RustBuffer.reserve(self.rbuf, numBytes)
+        yield None
+        self.rbuf.len += numBytes
+
+    def _pack_into(self, size, format, value):
+        with self._reserve(size):
+            # XXX TODO: I feel like I should be able to use `struct.pack_into` here but can't figure it out.
+            for i, byte in enumerate(struct.pack(format, value)):
+                self.rbuf.data[self.rbuf.len + i] = byte
+
+    def write(self, value):
+        with self._reserve(len(value)):
+            for i, byte in enumerate(value):
+                self.rbuf.data[self.rbuf.len + i] = byte
+
+class RustBufferStream(object):
+    """Helper for structured reading of values from a RustBuffer."""
+
+    def __init__(self, rbuf):
+        self.rbuf = rbuf
+        self.offset = 0
+
+    def remaining(self):
+        return self.rbuf.len - self.offset
+
+    def _unpack_from(self, size, format):
+        if self.offset + size > self.rbuf.len:
+            raise InternalError("read past end of rust buffer")
+        value = struct.unpack(format, self.rbuf.data[self.offset:self.offset+size])[0]
+        self.offset += size
+        return value
+
+    def read(self, size):
+        if self.offset + size > self.rbuf.len:
+            raise InternalError("read past end of rust buffer")
+        data = self.rbuf.data[self.offset:self.offset+size]
+        self.offset += size
+        return data
+
+
+
+class FFIConverterPrimitive:
+    size = NotImplemented
+    pack_fmt = NotImplemented
+
+
+    @staticmethod
+    def lower(builder, value):
+        builder._pack_into(self.size, self.pack_fmt, value)
+
+    @staticmethod
+    def lift(stream, value):
+        return stream._unpack_from(self.size, self.pack_fmt, value)
```

Comment:
This code is totally wrong, but hopefully you can imagine it calling `RustBuffer.consumerWithStream()` then unpacking from there.